README.md
# FAB Circular Menu 
[![Pub](https://img.shields.io/pub/v/fab_circular_menu.svg)](https://pub.dev/packages/fab_circular_menu)
[![Pull Requests are welcome](https://img.shields.io/badge/license-MIT-blue)](https://github.com/marianocordoba/fab-circular-menu/blob/master/LICENSE)
[![Codemagic build status](https://api.codemagic.io/apps/5cf6ad31434563000a9534d5/5cf6ad31434563000a9534d4/status_badge.svg)](https://codemagic.io/apps/5cf6ad31434563000a9534d5/5cf6ad31434563000a9534d4/latest_build)
![Null safety](https://img.shields.io/badge/null%20safety-true-brightgreen)

A Flutter package to create a nice circular menu using a Floating Action Button.

Inspired by [Mayur Kshirsagar](https://dribbble.com/mayurksgr)'s great [FAB Microinteraction](https://dribbble.com/shots/4354100-Daily-UI-Challenge-Day-75-FAB-Microinteraction) design.

![Showcase](https://i.imgur.com/ErrNnAw.gif)

## Installation

Just add `fab_circular_menu` to your [pubspec.yml](https://flutter.io/using-packages/) file

```yml
dependencies:
  fab_circular_menu: ^1.0.0
```

## Example

```dart
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Placeholder(),
        floatingActionButton: FabCircularMenu(
          children: <Widget>[
            IconButton(icon: Icon(Icons.home), onPressed: () {
              print('Home');
            }),
            IconButton(icon: Icon(Icons.favorite), onPressed: () {
              print('Favorite');
            })
          ]
        )
      )
    );
  }
}
```

You can check for a more complete example in the [example](https://github.com/marianocordoba/fab-circular-menu/tree/master/example) directory.

## Customize

You can customize the widget appareance using the following properties:

| Property  | Description | Default |
|----------|-------------|---------|
| alignment | Sets the widget alignment | `Alignment.bottomRight` |
| ringColor | Sets the ring color | `accentColor` |
| ringDiameter | Sets the ring diameter | `screenWidth * 1.25` (portrait) <br> `screenHeight * 1.25` (landscape) |
| ringWidth | Sets the ring width | `ringDiameter * 0.3` |
| fabSize | Sets the FAB size | `64.0` |
| fabElevation | Sets the elevation for the FAB | `8.0` |
| fabColor | Sets the FAB color | `primaryColor` |
| fabOpenColor | Sets the FAB color while the menu is open. This property takes precedence over `fabColor` | - |
| fabCloseColor | Sets the FAB color while the menu is closed. This property takes precedence over `fabColor` | - |
| fabChild | Sets the child inside the FAB. This property takes precedence over `fabOpenicon` and `fabCloseIcon` | - |
| fabOpenIcon | Sets the FAB icon while the menu is open | `Icon(Icons.menu)` |
| fabCloseIcon | Sets the FAB icon while the menu is closed | `Icon(Icons.close)` |
| fabMargin | Sets the widget margin | `EdgeInsets.all(16.0)` |
| animationDuration | Changes the animation duration | `Duration(milliseconds: 800)` |
| animationCurve | Allows you to modify the animation curve | `Curves.easeInOutCirc` |
| onDisplayChange | This callback is called every time the menu is opened or closed, passing the current state as a parameter. | - |

## Handling the menu programmatically

It is possible to handle the menu programatically by using a [key](https://api.flutter.dev/flutter/foundation/Key-class.html). Just create a key and set it in the `key` property of the `FabCircularMenu`, then use the key to get the current state and open, close or check the status of the menu.

```dart
class MyApp extends StatelessWidget {
  final GlobalKey<FabCircularMenuState> fabKey = GlobalKey();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: RaisedButton(
          onPressed: () {
            if (fabKey.currentState.isOpen) {
              fabKey.currentState.close();
            } else {
              fabKey.currentState.open();
            }
          },
          child: Text('Toggle menu')
        ),
        floatingActionButton: FabCircularMenu(
          key: fabKey,
          children: <Widget>[
            // ...
          ]
        )
      )
    );
  }
}
```

## Contributing

I will be very happy if you contribute to this project, please submit a PR 😁


CHANGELOG.md
## 1.1.0

 - Graduate package to a stable release. See pre-releases prior to this version for changelog entries.

## 1.1.0-dev.6

 - **FEAT**: Cleaning up for melos publish.
 - **FEAT**: bumping version.

## 1.1.0-dev.5

 - **FEAT**: Cleaning up for melos publish.
 - **FEAT**: bumping version.

## 1.1.0-dev.4

 - **FEAT**: Cleaning up for melos publish.
 - **FEAT**: bumping version.

## 1.1.0-dev.3

 - **FEAT**: Cleaning up for melos publish.
 - **FEAT**: bumping version.

## 1.1.0-dev.2

 - **FEAT**: Cleaning up for melos publish.
 - **FEAT**: bumping version.

## 1.1.0-dev.1

 - **FEAT**: Cleaning up for melos publish.
 - **FEAT**: bumping version.

## 1.1.0-dev.0

 - **FEAT**: bumping version.

## 1.0.2
* Null safety
* Changed the children count requirement to 1 from 2

## 1.0.1
* Fixes bug with alignment
* Added the posibility to modify the FAB border

## 1.0.0

* Added `alignment` property, allowing to align the FAB in the screen (See #5)
* Fixes animation errors in landscape orientation (See #9)
* Added the ability to access the state to open/close the menu programatically

#### **BREAKING**
* Removed `child` property
* `options` renamed to `children`
* Removed `controller` property
* `onDisplayChanged` signature changed to `void Function (bool isOpen)`

## 0.1.5

* #4 Added FabCircularMenuController to open/close the menu programatically

## 0.1.4

* #7 Add Close Color option and open/close change callback

## 0.1.3

* Bugfixes
* Improved animation

## 0.1.2

* Fixes multiple tickers error

## 0.1.1

* Added an example

## 0.1.0

* First functional release.


test/fab_circular_menu_test.dart
import 'package:fab_circular_menu/fab_circular_menu.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets("Should show it's options while opened",
      (WidgetTester tester) async {
    final widgetUnderTest = MaterialApp(
      home: Scaffold(
        floatingActionButton: FabCircularMenu(
          alignment: Alignment.bottomCenter,
          animationDuration: Duration.zero,
          children: const <Widget>[Icon(Icons.add), Icon(Icons.add)],
        ),
      ),
    );
    await tester.pumpWidget(widgetUnderTest);

    expect(find.byIcon(Icons.menu), findsOneWidget);
    expect(find.byIcon(Icons.close), findsNothing);
    expect(find.byIcon(Icons.add), findsNothing);

    await tester.tap(find.byType(RawMaterialButton));
    await tester.pump();

    expect(find.byIcon(Icons.menu), findsNothing);
    expect(find.byIcon(Icons.close), findsOneWidget);
    expect(find.byIcon(Icons.add), findsNWidgets(2));
  });

  testWidgets(
      'Should be able to be opened by tapping the FAB or programatically',
      (WidgetTester tester) async {
    final GlobalKey<FabCircularMenuState> fab = GlobalKey();
    bool isOpen = false;

    final widgetUnderTest = MaterialApp(
      home: Scaffold(
        floatingActionButton: FabCircularMenu(
          key: fab,
          alignment: Alignment.centerRight,
          animationDuration: Duration.zero,
          onDisplayChange: (val) {
            isOpen = val;
          },
          children: const <Widget>[Placeholder(), Placeholder()],
        ),
      ),
    );
    await tester.pumpWidget(widgetUnderTest);

    expect(isOpen, false);

    await tester.tap(find.byType(RawMaterialButton));
    await tester.pump();

    expect(isOpen, true);

    await tester.tap(find.byType(RawMaterialButton));
    await tester.pump();

    expect(isOpen, false);

    fab.currentState!.open();
    await tester.pump();

    expect(isOpen, true);
    expect(fab.currentState!.isOpen, true);

    fab.currentState!.close();
    await tester.pump();

    expect(isOpen, false);
    expect(fab.currentState!.isOpen, false);
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:fab_circular_menu/fab_circular_menu.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  final GlobalKey<FabCircularMenuState> fabKey = GlobalKey();

  @override
  Widget build(BuildContext context) {
    final primaryColor = Theme.of(context).primaryColor;

    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Container(
          color: const Color(0xFF192A56),
          child: Center(
            child: ElevatedButton(
              onPressed: () {
                // The menu can be handled programatically using a key
                if (fabKey.currentState.isOpen) {
                  fabKey.currentState.close();
                } else {
                  fabKey.currentState.open();
                }
              },
              color: Colors.white,
              child: Text('Toggle menu programatically', style: TextStyle(color: primaryColor)),
            ),
          ),
        ),
        floatingActionButton: Builder(
          builder: (context) => FabCircularMenu(
            key: fabKey,
            // Cannot be `Alignment.center`
            alignment: Alignment.bottomRight,
            ringColor: Colors.white.withAlpha(25),
            ringDiameter: 500.0,
            ringWidth: 150.0,
            fabSize: 64.0,
            fabElevation: 8.0,
            fabIconBorder: CircleBorder(),
            // Also can use specific color based on wether
            // the menu is open or not:
            // fabOpenColor: Colors.white
            // fabCloseColor: Colors.white
            // These properties take precedence over fabColor
            fabColor: Colors.white,
            fabOpenIcon: Icon(Icons.menu, color: primaryColor),
            fabCloseIcon: Icon(Icons.close, color: primaryColor),
            fabMargin: const EdgeInsets.all(16.0),
            animationDuration: const Duration(milliseconds: 800),
            animationCurve: Curves.easeInOutCirc,
            onDisplayChange: (isOpen) {
              _showSnackBar(context, "The menu is ${isOpen ? "open" : "closed"}");
            },
            children: <Widget>[
              RawMaterialButton(
                onPressed: () {
                  _showSnackBar(context, "You pressed 1");
                },
                shape: CircleBorder(),
                padding: const EdgeInsets.all(24.0),
                child: Icon(Icons.looks_one, color: Colors.white),
              ),
              RawMaterialButton(
                onPressed: () {
                  _showSnackBar(context, "You pressed 2");
                },
                shape: CircleBorder(),
                padding: const EdgeInsets.all(24.0),
                child: Icon(Icons.looks_two, color: Colors.white),
              ),
              RawMaterialButton(
                onPressed: () {
                  _showSnackBar(context, "You pressed 3");
                },
                shape: CircleBorder(),
                padding: const EdgeInsets.all(24.0),
                child: Icon(Icons.looks_3, color: Colors.white),
              ),
              RawMaterialButton(
                onPressed: () {
                  _showSnackBar(context, "You pressed 4. This one closes the menu on tap");
                  fabKey.currentState.close();
                },
                shape: CircleBorder(),
                padding: const EdgeInsets.all(24.0),
                child: Icon(Icons.looks_4, color: Colors.white),
              )
            ],
          ),
        ),
      ),
    );
  }

  void _showSnackBar(BuildContext context, String message) {
    Scaffold.of(context).showSnackBar(SnackBar(
      content: Text(message),
      duration: const Duration(milliseconds: 1000),
    ));
  }
}


lib/stack_with_all_children_receive_events.dart
import 'package:flutter/widgets.dart';
import 'package:flutter/rendering.dart';

/// Passes all events to all children of the stack.  The FAB was having issues
/// where the padding on the button was blocking the circular items on the edge
class StackWithAllChildrenReceiveEvents extends Stack {
  StackWithAllChildrenReceiveEvents({
    Key? key,
    AlignmentGeometry alignment = AlignmentDirectional.topStart,
    TextDirection textDirection = TextDirection.ltr,
    StackFit fit = StackFit.loose,
    Overflow overflow = Overflow.clip,
    List<Widget> children = const <Widget>[],
    Clip clipBehavior = Clip.hardEdge,
  }) : super(
          key: key,
          alignment: alignment,
          textDirection: textDirection,
          fit: fit,
          overflow: overflow,
          clipBehavior: clipBehavior,
          children: children,
        );

  @override
  RenderStackWithAllChildrenReceiveEvents createRenderObject(
      BuildContext context) {
    return RenderStackWithAllChildrenReceiveEvents(
      alignment: alignment,
      textDirection: textDirection ?? Directionality.of(context),
      fit: fit,
      clipBehavior: clipBehavior == Overflow.visible ? Clip.none : clipBehavior,
    );
    // return RenderStackWithAllChildrenReceiveEvents(
    //   alignment: alignment,
    //   textDirection: textDirection ?? Directionality.of(context),
    //   fit: fit,
    //   overflow: overflow,
    // );
  }

  @override
  void updateRenderObject(BuildContext context,
      RenderStackWithAllChildrenReceiveEvents renderObject) {
    renderObject
      ..alignment = alignment
      ..textDirection = textDirection ?? Directionality.of(context)
      ..fit = fit
      ..clipBehavior =
          clipBehavior == Overflow.visible ? Clip.none : clipBehavior;
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties
        .add(DiagnosticsProperty<AlignmentGeometry>('alignment', alignment));
    properties.add(EnumProperty<TextDirection>('textDirection', textDirection,
        defaultValue: null));
    properties.add(EnumProperty<StackFit>('fit', fit));
    properties.add(EnumProperty<Overflow>('overflow', overflow));
    properties.add(EnumProperty<Clip>('clipBehavior', clipBehavior));
  }
}

class RenderStackWithAllChildrenReceiveEvents extends RenderStack {
  RenderStackWithAllChildrenReceiveEvents({
    List<RenderBox> children = const [],
    AlignmentGeometry alignment = AlignmentDirectional.topStart,
    TextDirection? textDirection,
    StackFit fit = StackFit.loose,
    Clip clipBehavior = Clip.hardEdge,
  }) : super(
          alignment: alignment,
          textDirection: textDirection,
          fit: fit,
          clipBehavior: clipBehavior,
        );

  @override
  bool defaultHitTestChildren(HitTestResult result,
      {required Offset position}) {
    // the x, y parameters have the top left of the node's box as the origin
    var child = lastChild;
    while (child != null) {
      StackParentData childParentData = child.parentData as StackParentData;
      child.hitTest(result as BoxHitTestResult,
          position: position - childParentData.offset);
      child = childParentData.previousSibling;
    }
    return false;
  }

  @override
  bool hitTestChildren(HitTestResult result, {required Offset position}) {
    return defaultHitTestChildren(result, position: position);
  }
}


lib/fab_circular_menu.dart
import 'dart:math';

import 'package:fab_circular_menu/stack_with_all_children_receive_events.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:vector_math/vector_math.dart' as vector;

typedef DisplayChange = void Function(bool isOpen);

class FabCircularMenu extends StatefulWidget {
  /// When used as the floatingActionButton in a Scaffold, there is a default margin
  /// applied that needs to be removed.  If this button is placed manually, this offset
  /// can cause issues.  Set this value to false to _not_ attempt to remove the default
  /// FAB margin
  final bool removeDefaultFabMargin;

  final List<Widget> children;
  final Alignment alignment;
  final Color? ringColor;
  final double? ringDiameter;
  final double? ringWidth;
  final Key? buttonKey;
  final double rangeStartOffset;
  final double rangeEndOffset;
  final double fabSize;
  final double fabElevation;
  final Color? fabColor;
  final Color? fabOpenColor;
  final Color? fabCloseColor;
  final Widget? fabChild;
  final Widget fabOpenIcon;
  final Widget fabCloseIcon;
  final ShapeBorder? fabIconBorder;
  final EdgeInsets fabMargin;
  final Offset menuOffset;
  final Duration animationDuration;
  final Curve animationCurve;
  final DisplayChange? onDisplayChange;

  /// Disables the menu from appearing, in favor of a simple onTap handler
  final VoidCallback? onFabTap;

  FabCircularMenu(
      {Key? key,
      this.buttonKey,
      this.alignment = Alignment.bottomCenter,
      this.ringColor,
      this.ringDiameter,
      this.ringWidth,
      this.onFabTap,
      this.fabSize = 64.0,
      this.menuOffset = Offset.zero,
      this.fabElevation = 8.0,
      this.fabColor,
      this.fabOpenColor,
      this.removeDefaultFabMargin = true,
      this.fabCloseColor,
      this.fabIconBorder,
      this.rangeStartOffset = 0.0,
      this.rangeEndOffset = 0.0,
      this.fabChild,
      this.fabOpenIcon = const Icon(Icons.menu),
      this.fabCloseIcon = const Icon(Icons.close),
      this.fabMargin = const EdgeInsets.all(16.0),
      this.animationDuration = const Duration(milliseconds: 800),
      this.animationCurve = Curves.easeInOutCirc,
      this.onDisplayChange,
      required this.children})
      : assert(children.length >= 1),
        super(key: key);

  @override
  FabCircularMenuState createState() => FabCircularMenuState();
}

class FabCircularMenuState extends State<FabCircularMenu>
    with SingleTickerProviderStateMixin {
  late double _screenWidth;
  late double _screenHeight;
  late double _marginH;
  late double _marginV;
  late double _directionX;
  late double _directionY;
  late double _translationX;
  late double _translationY;

  Color? _ringColor;
  double? _ringDiameter;
  double? _ringWidth;
  Color? _fabColor;
  Color? _fabOpenColor;
  Color? _fabCloseColor;
  late ShapeBorder _fabIconBorder;

  late AnimationController _animationController;
  late Animation<double> _scaleAnimation;
  late Animation _scaleCurve;
  late Animation<double> _rotateAnimation;
  late Animation _rotateCurve;
  Animation<Color?>? _colorAnimation;
  late Animation _colorCurve;

  bool _isOpen = false;
  bool _isAnimating = false;

  @override
  void initState() {
    super.initState();

    _animationController =
        AnimationController(duration: widget.animationDuration, vsync: this);

    _scaleCurve = CurvedAnimation(
        parent: _animationController,
        curve: Interval(0.0, 0.4, curve: widget.animationCurve));
    _scaleAnimation = Tween<double>(begin: 0.0, end: 1.0)
        .animate(_scaleCurve as Animation<double>)
      ..addListener(() {
        setState(() {});
      });

    _rotateCurve = CurvedAnimation(
        parent: _animationController,
        curve: Interval(0.4, 1.0, curve: widget.animationCurve));
    _rotateAnimation = Tween<double>(begin: 0.5, end: 1.0)
        .animate(_rotateCurve as Animation<double>)
      ..addListener(() {
        setState(() {});
      });
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _calculateProps();
  }

  @override
  Widget build(BuildContext context) {
    // This makes the widget able to correctly redraw on
    // hot reload while keeping performance in production
    if (!kReleaseMode) {
      _calculateProps();
    }

    return Container(
      // Removes the default FAB margin
      transform: widget.removeDefaultFabMargin
          ? Matrix4.translationValues(16.0, 16.0, 0.0)
          : null,
      child: StackWithAllChildrenReceiveEvents(
        overflow: Overflow.visible,
        alignment: widget.alignment,
        children: <Widget>[
          // Ring
          Transform(
            transform: Matrix4.translationValues(
              _translationX,
              _translationY,
              0.0,
            )..scale(_scaleAnimation.value),
            alignment: FractionalOffset.center,
            child: OverflowBox(
              maxWidth: _ringDiameter,
              maxHeight: _ringDiameter,
              child: Container(
                width: _ringDiameter,
                height: _ringDiameter,
                child: CustomPaint(
                  painter: _RingPainter(
                    width: _ringWidth,
                    color: _ringColor,
                  ),
                  child: _scaleAnimation.value == 1.0
                      ? Transform.rotate(
                          angle: (2 * pi) *
                              _rotateAnimation.value *
                              _directionX *
                              _directionY,
                          child: Container(
                            child: Stack(
                              alignment: Alignment.center,
                              children: widget.children
                                  .asMap()
                                  .map((index, child) => MapEntry(index,
                                      _applyTransformations(child, index)))
                                  .values
                                  .toList(),
                            ),
                          ),
                        )
                      : Container(),
                ),
              ),
            ),
          ),

          // FAB
          Container(
            margin: widget.fabMargin,
            width: widget.fabSize,
            height: widget.fabSize,
            child: RawMaterialButton(
              key: widget.buttonKey,
              fillColor: _colorAnimation!.value,
              shape: _fabIconBorder,
              elevation: widget.fabElevation,
              onPressed: () {
                if (widget.onFabTap != null) {
                  widget.onFabTap!();
                  return;
                } else {
                  if (_isAnimating) return;

                  if (_isOpen) {
                    close();
                  } else {
                    open();
                  }
                }
              },
              child: Center(
                child: widget.fabChild == null
                    ? (_scaleAnimation.value == 1.0
                        ? widget.fabCloseIcon
                        : widget.fabOpenIcon)
                    : widget.fabChild,
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _applyTransformations(Widget child, int index) {
    double angleFix = 0.0;
    if (widget.alignment.x == 0) {
      angleFix = 45.0 * _directionY.abs();
    } else if (widget.alignment.y == 0) {
      angleFix = -45.0 * _directionX.abs();
    }
    angleFix += widget.rangeStartOffset;

    final angle = vector.radians(
        (90.0 - widget.rangeStartOffset - widget.rangeEndOffset) /
                (widget.children.length - 1) *
                index +
            angleFix);

    return Transform(
        transform: Matrix4.translationValues(
            (-(_ringDiameter! / 2) * cos(angle) +
                    (_ringWidth! / 2 * cos(angle))) *
                _directionX,
            (-(_ringDiameter! / 2) * sin(angle) +
                    (_ringWidth! / 2 * sin(angle))) *
                _directionY,
            0.0),
        alignment: FractionalOffset.center,
        child: Material(
          color: Colors.transparent,
          child: child,
        ));
  }

  void _calculateProps() {
    _ringColor = widget.ringColor ?? Theme.of(context).colorScheme.secondary;
    _fabColor = widget.fabColor ?? Theme.of(context).primaryColor;
    _fabOpenColor = widget.fabOpenColor ?? _fabColor;
    _fabCloseColor = widget.fabCloseColor ?? _fabColor;
    _fabIconBorder = widget.fabIconBorder ?? CircleBorder();
    _screenWidth = MediaQuery.of(context).size.width;
    _screenHeight = MediaQuery.of(context).size.height;
    _ringDiameter =
        widget.ringDiameter ?? min(_screenWidth, _screenHeight) * 1.25;
    _ringWidth = widget.ringWidth ?? _ringDiameter! * 0.3;
    _marginH = (widget.fabMargin.right + widget.fabMargin.left) / 2;
    _marginV = (widget.fabMargin.top + widget.fabMargin.bottom) / 2;
    _directionX = widget.alignment.x == 0 ? 1 : 1 * widget.alignment.x.sign;
    _directionY = widget.alignment.y == 0 ? 1 : 1 * widget.alignment.y.sign;
    _translationX = (((_screenWidth - widget.fabSize) / 2 - _marginH) *
            widget.alignment.x) +
        widget.menuOffset.dx;
    _translationY = (((_screenHeight - widget.fabSize) / 2 - _marginV) *
            widget.alignment.y) +
        widget.menuOffset.dy;

    if (_colorAnimation == null || !kReleaseMode) {
      _colorCurve = CurvedAnimation(
          parent: _animationController,
          curve: Interval(
            0.0,
            0.4,
            curve: widget.animationCurve,
          ));
      _colorAnimation = ColorTween(begin: _fabCloseColor, end: _fabOpenColor)
          .animate(_colorCurve as Animation<double>)
        ..addListener(() {
          setState(() {});
        });
    }
  }

  void open() {
    _isAnimating = true;
    _animationController.forward().then((_) {
      _isAnimating = false;
      _isOpen = true;
      if (widget.onDisplayChange != null) {
        widget.onDisplayChange!(true);
      }
    });
  }

  void close() {
    _isAnimating = true;
    _animationController.reverse().then((_) {
      _isAnimating = false;
      _isOpen = false;
      if (widget.onDisplayChange != null) {
        widget.onDisplayChange!(false);
      }
    });
  }

  bool get isOpen => _isOpen;
}

class _RingPainter extends CustomPainter {
  final double? width;
  final Color? color;

  _RingPainter({
    required this.width,
    this.color,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()
      ..color = color ?? Colors.white
      ..style = PaintingStyle.stroke
      ..strokeWidth = size.width < width! ? size.width : width!;

    canvas.drawArc(
      Rect.fromLTWH(
          width! / 2, width! / 2, size.width - width!, size.height - width!),
      0.0,
      2 * pi,
      false,
      paint,
    );
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}
